File: app/src/main/java/com/waqasout/AlarmReceiver.java
package com.waqasout;

import android.app.admin.DevicePolicyManager;
import android.content.BroadcastReceiver;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.widget.Toast;

public class AlarmReceiver extends BroadcastReceiver {

    @Override
    public void onReceive(Context context, Intent intent) {
        DevicePolicyManager devicePolicyManager = (DevicePolicyManager) context.getSystemService(Context.DEVICE_POLICY_SERVICE);
        ComponentName compName = new ComponentName(context, DeviceAdmin.class);

        if (devicePolicyManager.isAdminActive(compName)) {
            String action = intent.getAction();
            if ("com.waqasout.ACTION_LOCK_SCREEN".equals(action)) {
                devicePolicyManager.lockNow();
                Toast.makeText(context, "Scheduled Lock Activated", Toast.LENGTH_SHORT).show();
                Intent serviceIntent = new Intent(context, LockService.class);
                serviceIntent.setAction("com.waqasout.ACTION_LOCK_SCREEN");
                context.startService(serviceIntent);
            } else if ("com.waqasout.ACTION_UNLOCK_SCREEN".equals(action)) {
                Toast.makeText(context, "Scheduled Lock Deactivated. You can now unlock your device.", Toast.LENGTH_LONG).show();
                Intent serviceIntent = new Intent(context, LockService.class);
                serviceIntent.setAction("com.waqasout.ACTION_UNLOCK_SCREEN");
                context.startService(serviceIntent);
            }
        } else {
            Toast.makeText(context, "Device Admin not active. Cannot perform scheduled lock/unlock.", Toast.LENGTH_LONG).show();
        }
    }
}

File: android/app/src/main/kotlin/com/screenlockapp/screen_lock_app/AlarmReceiver.kt
package com.screenlockapp.screen_lock_app

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.util.Log

/**
 * Receiver for handling scheduled lock/unlock alarms
 */
class AlarmReceiver : BroadcastReceiver() {
    
    companion object {
        private const val TAG = "AlarmReceiver"
        const val ACTION_LOCK_SCREEN = "com.screenlockapp.ACTION_LOCK_SCREEN"
        const val ACTION_UNLOCK_SCREEN = "com.screenlockapp.ACTION_UNLOCK_SCREEN"
        const val EXTRA_SCHEDULE_ID = "schedule_id"
    }

    override fun onReceive(context: Context, intent: Intent) {
        val action = intent.action
        Log.d(TAG, "Received action: $action")

        val deviceAdminManager = DeviceAdminManager(context)
        
        when (action) {
            ACTION_LOCK_SCREEN -> handleLockScreen(context, intent, deviceAdminManager)
            ACTION_UNLOCK_SCREEN -> handleUnlockScreen(context, intent, deviceAdminManager)
        }
    }

    private fun handleLockScreen(context: Context, intent: Intent, adminManager: DeviceAdminManager) {
        Log.i(TAG, "Handling scheduled lock screen")
        
        if (adminManager.isAdminActive()) {
            val success = adminManager.lockScreen()
            if (success) {
                // Show notification that screen is locked
                NotificationHelper.showLockNotification(context)
                
                // Schedule unlock if duration is specified
                val scheduleId = intent.getIntExtra(EXTRA_SCHEDULE_ID, -1)
                if (scheduleId != -1) {
                    val scheduleManager = ScheduleManager(context)
                    scheduleManager.scheduleUnlockForActiveSchedule(scheduleId)
                }
            }
        } else {
            Log.w(TAG, "Device admin not active, cannot lock screen")
        }
    }

    private fun handleUnlockScreen(context: Context, intent: Intent, adminManager: DeviceAdminManager) {
        Log.i(TAG, "Handling scheduled unlock screen")
        
        if (adminManager.isAdminActive()) {
            val success = adminManager.unlockScreen()
            if (success) {
                // Show notification that screen is unlocked
                NotificationHelper.showUnlockNotification(context)
            }
        } else {
            Log.w(TAG, "Device admin not active, cannot unlock screen")
        }
    }
}

